functions/src/requests/index.ts

```diff
@@ -1,125 +1,147 @@
+  // We have a new CAV rating -  Update CAV rating average but only this time.
+  if (requestBefore?.cavRating === null && requestAfter?.cavRating !== null && requestAfter?.cavUserRef) {
+    const user = (await requestAfter.cavUserRef.withConverter(UserFirestoreConverter).get()).data();
+    if (user) {
+      const currentAverage = user.averageRating ? user.averageRating : 0;
+      const cavRatingsReceived = user.cavRatingsReceived ? user.cavRatingsReceived + 1 : 1;
+      // Calculate new average by summing all past averages and re-averaging with new value
+      const newAverage = ((currentAverage * cavRatingsReceived) + requestAfter.cavRating) / cavRatingsReceived;
+      operations.push(requestAfter.cavUserRef.update({
+        averageRating: newAverage,
+        cavRatingsReceived,
+      }));
+    }
+  } else if (
+    requestBefore?.cavRating !== null &&
+    requestAfter?.cavRating !== null &&
+    requestBefore?.cavRating !== requestAfter?.cavRating &&
+    requestBefore?.cavRatedAt &&
+    requestAfter?.cavRatedAt &&
+    requestBefore?.cavRatedAt !== requestAfter?.cavRatedAt &&
+    requestAfter?.cavUserRef
+  ) {
+    const previousRatedAt = moment(requestBefore.cavRatedAt.toDate());
+    const fiveMinutesPastPreviousRatedAt = previousRatedAt.add(5, 'minutes');
+    const currentRatedAt = moment(requestAfter.cavRatedAt.toDate());
+    if (currentRatedAt.isSameOrBefore(fiveMinutesPastPreviousRatedAt)) {
+      const user = (await requestAfter.cavUserRef.withConverter(UserFirestoreConverter).get()).data();
+      if (user) {
+        let average = user.averageRating ? user.averageRating : 0;
+        const cavRatingsReceived = user.cavRatingsReceived ? user.cavRatingsReceived : 2;
+        // Subtract the old rating from the average and reverse the average to what it was before the old rating was given
+        average = ((average * cavRatingsReceived) - requestBefore.cavRating) / (cavRatingsReceived - 1);
+        // Add the new rating to the average and calculate the new average
+        const newAverage = ((average * cavRatingsReceived) + requestAfter.cavRating) / cavRatingsReceived;
+        operations.push(requestAfter.cavUserRef.update({
+          averageRating: newAverage,
+        }));
+      }
+    }
+  }
+};
 
+const attemptToUpdateCavCompletedOffersCounts = (operations: Promise<void>[], requestBefore: Request | null, requestAfter: Request | null) => {
   // A request has just been completed - Update CAV request completed count
   if (
     requestBefore?.status !== RequestStatus.completed &&
     requestAfter?.status === RequestStatus.completed &&
     requestAfter.cavUserRef
   ) {
-    let user = UserFirestoreConverter.fromFirestore((await requestAfter.cavUserRef.get()));
+    // Using increment without read
     operations.push(requestAfter.cavUserRef.update({
-      casesCompleted: user.casesCompleted + 1
+      casesCompleted: admin.firestore.FieldValue.increment(1),
```

Comment:
An interesting firebase feature that lets us increment values without having to consume a read. It also works in a much safer fashion. `read before write` operations like the one this replaces are unsafe (unless conducted inside a transaction).

Reply:
oh. this is great. Thanks!